// @arco-design/web-vue/lib/checkbox/checkbox-group.js
"use strict";
var vue = require("vue");
var globalConfig = require("../_utils/global-config.js");
var context = require("./context.js");
var useFormItem = require("../_hooks/use-form-item.js");
var checkbox = require("./checkbox.js");
var is = require("../_utils/is.js");
var CheckboxGroup = vue.defineComponent({
  name: "CheckboxGroup",
  props: {
    modelValue: {
      type: Array,
      default: void 0
    },
    defaultValue: {
      type: Array,
      default: () => []
    },
    options: {
      type: Array
    },
    direction: {
      type: String,
      default: "horizontal"
    },
    disabled: {
      type: Boolean,
      default: false
    }
  },
  emits: {
    "update:modelValue": (value) => true,
    "change": (value, ev) => true
  },
  setup(props, {
    emit,
    slots
  }) {
    const {
      disabled
    } = vue.toRefs(props);
    const prefixCls = globalConfig.getPrefixCls("checkbox-group");
    const {
      mergedDisabled,
      eventHandlers
    } = useFormItem.useFormItem({
      disabled
    });
    const _value = vue.ref(props.defaultValue);
    const computedValue = vue.computed(() => {
      var _a;
      return (_a = props.modelValue) != null ? _a : _value.value;
    });
    const options = vue.computed(() => {
      var _a;
      return ((_a = props.options) != null ? _a : []).map((option) => {
        if (is.isString(option) || is.isNumber(option)) {
          return {
            label: option,
            value: option
          };
        }
        return option;
      });
    });
    const handleChange = (value, e) => {
      var _a, _b;
      _value.value = value;
      emit("update:modelValue", value);
      emit("change", value, e);
      (_b = (_a = eventHandlers.value) == null ? void 0 : _a.onChange) == null ? void 0 : _b.call(_a, e);
    };
    vue.provide(context.checkboxGroupKey, vue.reactive({
      name: "ArcoCheckboxGroup",
      computedValue,
      disabled: mergedDisabled,
      slots,
      handleChange
    }));
    const cls = vue.computed(() => [prefixCls, `${prefixCls}-direction-${props.direction}`]);
    vue.watch(() => props.modelValue, (curValue) => {
      if (curValue) {
        _value.value = [...curValue];
      }
    });
    const renderOptions = () => {
      return options.value.map((option, index) => vue.createVNode(checkbox, {
        "key": option.value,
        "value": option.value,
        "disabled": option.disabled,
        "indeterminate": option.indeterminate,
        "modelValue": computedValue.value.includes(option.value)
      }, {
        default: () => [slots.label ? slots.label({
          data: option
        }) : is.isFunction(option.label) ? option.label() : option.label]
      }));
    };
    return () => {
      var _a;
      return vue.createVNode("span", {
        "class": cls.value
      }, [options.value.length > 0 ? renderOptions() : (_a = slots.default) == null ? void 0 : _a.call(slots)]);
    };
  }
});
module.exports = CheckboxGroup;
