// @arco-design/web-vue/es/_components/virtual-list/hooks/use-item-height.js
import { toRefs, ref, computed, watch } from "vue";
const DEFAULT_ITEM_HEIGHT = 32;
function useItemHeight(props) {
  const { estimatedItemHeight: propEstimatedItemHeight, data } = toRefs(props);
  const itemHeightCacheMap = ref(new Map());
  const estimatedItemHeight = ref(propEstimatedItemHeight == null ? void 0 : propEstimatedItemHeight.value);
  const itemLength = computed(() => itemHeightCacheMap.value.size);
  watch(itemLength, () => {
    if (itemLength.value && !estimatedItemHeight.value) {
      estimatedItemHeight.value = [...itemHeightCacheMap.value.entries()].reduce((sum, [, height]) => sum + height, 0) / itemLength.value;
    }
  });
  const itemHeight = computed(() => estimatedItemHeight.value || DEFAULT_ITEM_HEIGHT);
  const minItemHeight = computed(() => Math.min(itemHeight.value, DEFAULT_ITEM_HEIGHT));
  const totalHeight = computed(() => itemHeight.value * data.value.length);
  function setItemHeight(key, height) {
    itemHeightCacheMap.value.set(key, height);
  }
  function getItemHeight(key) {
    return itemHeightCacheMap.value.get(key);
  }
  function getItemHeightOrDefault(key) {
    return itemHeightCacheMap.value.get(key) || itemHeight.value;
  }
  function getItemHeightByIndex(index) {
    const { key } = data.value[index];
    return itemHeightCacheMap.value.get(key);
  }
  function getItemHeightOrDefaultByIndex(index) {
    return getItemHeightByIndex(index) || itemHeight.value;
  }
  return {
    itemHeight,
    minItemHeight,
    estimatedItemHeight,
    totalHeight,
    setItemHeight,
    getItemHeight,
    getItemHeightOrDefault,
    getItemHeightByIndex,
    getItemHeightOrDefaultByIndex
  };
}
export { useItemHeight };
