// @arco-design/web-vue/lib/tree/node-switcher.vue_vue&type=script&lang.js
"use strict";
var vue = require("vue");
var renderFunction = require("../_components/render-function.js");
var index = require("../icon/icon-loading/index.js");
var iconHover = require("../_components/icon-hover.js");
var index$1 = require("../icon/icon-caret-down/index.js");
var index$2 = require("../icon/icon-file/index.js");
var useTreeContext = require("./hooks/use-tree-context.js");
var usePickSlots = require("../_hooks/use-pick-slots.js");
var _sfc_main = vue.defineComponent({
  name: "TreeNodeSwitcher",
  components: {
    IconLoading: index,
    RenderFunction: renderFunction
  },
  props: {
    prefixCls: String,
    expanded: Boolean,
    loading: Boolean,
    showLine: Boolean,
    isLeaf: Boolean,
    treeNodeData: {
      type: Object
    },
    icons: {
      type: Object
    }
  },
  emits: ["click"],
  setup(props, {
    slots,
    emit
  }) {
    const {
      icons
    } = vue.toRefs(props);
    const treeContext = useTreeContext();
    const nodeSwitcherIcon = usePickSlots(slots, "switcher-icon");
    const nodeLoadingIcon = usePickSlots(slots, "loading-icon");
    return {
      switcherIcon: vue.computed(() => {
        var _a;
        return ((_a = icons == null ? void 0 : icons.value) == null ? void 0 : _a.switcherIcon) || nodeSwitcherIcon.value || treeContext.switcherIcon;
      }),
      loadingIcon: vue.computed(() => {
        var _a;
        return ((_a = icons == null ? void 0 : icons.value) == null ? void 0 : _a.loadingIcon) || nodeLoadingIcon.value || treeContext.loadingIcon;
      }),
      onClick(e) {
        emit("click", e);
      }
    };
  },
  render() {
    const {
      prefixCls,
      switcherIcon,
      loadingIcon,
      onClick,
      expanded,
      loading,
      isLeaf,
      showLine,
      treeNodeData
    } = this;
    if (loading) {
      return loadingIcon ? loadingIcon(treeNodeData) : vue.h(index);
    }
    let icon = null;
    let needIconHover = false;
    if (!isLeaf) {
      const defaultIcon = showLine ? vue.h("span", {
        class: `${prefixCls}-${expanded ? "minus" : "plus"}-icon`
      }) : vue.h(index$1);
      icon = switcherIcon ? switcherIcon(treeNodeData) : defaultIcon;
      needIconHover = !showLine;
    } else if (showLine) {
      icon = switcherIcon ? switcherIcon(treeNodeData) : vue.h(index$2);
    }
    if (!icon)
      return null;
    const content = vue.h("span", {
      class: `${prefixCls}-switcher-icon`,
      onClick
    }, icon);
    return needIconHover ? vue.h(iconHover, {
      class: `${prefixCls}-icon-hover`
    }, () => content) : content;
  }
});
module.exports = _sfc_main;
